docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,134 @@
+
+Only three methods need to be implemented for a `RecipeSerializer`:
+
+Method | Description
+:---: | :---
+fromJson | Decodes a JSON into the `Recipe` subtype.
+toNetwork | Encodes a `Recipe` to the buffer to send to the client. The recipe identifier does not need to be encoded.
+fromNetwork | Decodes a `Recipe` from the buffer sent from the server. The recipe identifier does not need to be decoded.
+
+A `RecipeSerializer` must be [registered][forge].
+
+!!! tip
+    For ease of convenience, the `RecipeSerializer` subtype can extend `ForgeRegistryEntry` to implement the methods within `IForgeRegistryEntry`.
+
+```java
+public class ExampleSerializer extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<ExampleRecipe> {
+  // Implement methods here
+}
+```
+
+!!! tip
+    There are some useful methods to make reading and writing data for recipes easier. `Ingredient`s can use `#fromJson`, `#toNetwork`, and `#fromNetwork` while `ItemStack`s can use `CraftingHelper#getItemStack`, `FriendlyByteBuf#writeItem`, and `FriendlyByteBuf# readItem`.
+
+The `RecipeSerializer` instance must then be returned by `Recipe#getSerializer` in the new recipe subtype.
+
+```java
+// For some RecipeSerializer EXAMPLE_SERIALIZER
+// In ExampleRecipe
+@Override
+public RecipeSerializer<?> getSerializer() {
+  return EXAMPLE_SERIALIZER;
+}
+```
+
+Building the JSON
+-----------------
+
+Custom Recipe JSONs are stored in the same place as other [recipes][json]. The specified `type` should represent the registry name of the **recipe serializer**. Any additional data is specified by the serializer during decoding.
+
+```json5
+{
+  // The custom serializer registry name
+  "type": "examplemod:example_serializer",
+  "input": {
+    // Some ingredient input
+  },
+  "data": 0, // Some data wanted for the recipe
+  "output": {
+    // Some stack output
+  }
+}
+```
+
+Non-Item Logic
+--------------
+
+If items are not used as part of the input or result of a recipe, then the normal methods provided in [`RecipeManager`][manager] will not be useful. Instead, an additional method for testing a recipe's validity and/or supplying the result should be added to the custom `Recipe` instance. From there, all the recipes for that specific `RecipeType` can be obtained via `RecipeManager#getAllRecipesFor` and then checked and/or supplied the result using the newly implemented methods.
+
+```java
+// In some Recipe subimplementation ExampleRecipe
+
+// Checks the block at the position to see if it matches the stored data
+boolean matches(Level level, BlockPos pos);
+
+// Creates the block state to set the block at the specified position to
+BlockState assemble();
+
+// In some manager class
+public Optional<ExampleRecipe> getRecipeFor(Level level, BlockPos pos) {
+  return level.getRecipeManager().getAllRecipesFor(exampleRecipeType)
+    .stream().filter(recipe -> recipe.matches(level, pos)).findFirst();
```

Comment:
Perhaps for readability, each operation in this stream should be on its own separate line.